google/google.go

```diff
@@ -117,73 +131,130 @@ func (g *GoogleCloud) subscribe(opts ps.HandlerOptions, h ps.MsgHandler, ready c
-		// your service if you can't process a message
-		// and will also not handle
-		sem := semaphore.NewWeighted(int64(opts.Concurrency))
-
 		// Listen to messages and call the MsgHandler
 		for {
 			if g.shutdown {
 				break
 			}
 
-			cctx, cancel := context.WithCancel(context.Background())
-			err = sub.Receive(cctx, func(ctx context.Context, m *pubsub.Message) {
-				if serr := sem.Acquire(ctx, 1); serr != nil {
-					logrus.Errorf(
-						"pubsub: Failed to acquire worker semaphore: %v",
-						serr,
-					)
-					return
-				}
-				defer sem.Release(1)
-
-				b.Reset()
-				msg := ps.Msg{
-					ID:          m.ID,
-					Metadata:    m.Attributes,
-					Data:        m.Data,
-					PublishTime: &m.PublishTime,
-					Ack: func() {
-						m.Ack()
-					},
-					Nack: func() {
-						m.Nack()
-					},
-				}
-
-				err = h(ctx, msg)
-				if err != nil {
-					return
-				}
-
-				if opts.AutoAck {
-					m.Ack()
-				}
-			})
+			req := &pb.PullRequest{
+				Subscription: fmt.Sprintf("projects/%s/subscriptions/%s", g.projectID, subName),
+				MaxMessages:  int32(opts.Concurrency),
+			}
+
+			conn, err := g.defaultConn(context.Background())
+			if err != nil {
+				break
+			}
 
+			resp, err := conn.Pull(context.Background(), req)
 			if err != nil {
 				d := b.Duration()
 				logrus.Errorf(
-					"Subscription receive to topic %s failed, reconnecting in %v. Err: %v",
+					"pull from topic %s failed, retrying in %v. Err: %v",
 					opts.Topic, d, err,
 				)
 				time.Sleep(d)
+				continue
+			}
+
+			logr.WithCtx(context.Background()).Debugf(
+				"Google Pubsub: pulled %d messages for %s",
+				len(resp.ReceivedMessages),
+				subName,
+			)
+
+			b.Reset()
+
+			var wg sync.WaitGroup
+			for _, m := range resp.ReceivedMessages {
+				go func() {
+					wg.Add(1)
+					defer wg.Done()
+					done := make(chan struct{})
+
+					cctx, cancel := context.WithTimeout(context.Background(), opts.Deadline)
+					defer cancel()
+
+					go func(m *pb.ReceivedMessage) {
+						defer func() {
+							done <- struct{}{}
+						}()
+
+						t, _ := ptypes.Timestamp(m.Message.PublishTime)
+						msg := ps.Msg{
+							ID:          m.Message.MessageId,
+							Metadata:    m.Message.Attributes,
+							Data:        m.Message.Data,
+							PublishTime: &t,
+							Ack: func() {
+								g.AckMessage(cctx, subName, []string{m.AckId})
+							},
+							Nack: func() {
+								g.ModAckMessage(cctx, subName, []string{m.AckId}, 0)
+							},
+						}
+
+						err = h(cctx, msg)
+						if err != nil {
+							timeout := 5 * time.Minute
+							g.ModAckMessage(cctx, subName, []string{m.AckId}, int32(timeout))
+							return
+						}
+
+						if opts.AutoAck {
+							g.AckMessage(cctx, subName, []string{m.AckId})
+						}
+					}(m)
+
+					select {
+					case <-done:
+					case <-time.After(opts.Deadline):
+						logr.WithCtx(cctx).Debugf("Google Pubsub: worker timed out %v", subName)
+					}
+				}()
```

Comment:
`m` should be also bound to the outer goroutine, otherwise `m` might have the value of the last received message in the loop, for every received message.